### ISSUE TITLE
DNS over https (DOH)

### ISSUE THREAD

**Issue by robjob1938** [mullvad/mullvadvpn-app#4385]
## Issue report

Operating system: All

App version: 2023.1

### Issue description

I really appreciate Mullvad adding a custom DNS server option, it actually made me look at your service again instead of IVPN (I purchased 1 month to try it out).  Your service would be perfect (in my eyes) if it allowed either DOT or DOH (DNS over https and/or DNS over TLS).  This one option is the only thing that keeps me from fully implementing your service over IVPN (which allows a custom DNS server to DOH).

**Comment by MrChocolatine** [comment]
@robjob1938 , you might interested in this article
https://mullvad.net/en/help/dns-over-https-and-dns-over-tls/

**Comment by robjob1938** [comment]
> @robjob1938 , you might be interested in this article https://mullvad.net/en/help/dns-over-https-and-dns-over-tls/

Thanks MrChocolatine, I understand Mullvad's DNS,  I want to use my own DNS server via mullvad app to connect to any number of DNS services such as Quad9 https://www.quad9.net/service/service-addresses-and-features/

HTTPS:
https://dns.quad9.net/dns-query

TLS:
tls://dns.quad9.net

**Comment by MrChocolatine** [comment]
Sorry but I am not part of their team, I am just a happy customer that likes to help.

**Comment by Leprous1** [comment]
I agree, it would be awesome. I use custom DNS (Adguard) to block tracking and ads but I always had DNS leak since I cannot use DOH nor DOT which is a bummer.

Or Mullvad could add a way to blacklist custom domain instead of having premade lists but I assume this would be more work than permitting DOH.

**Comment by pinkisemils** [comment]
The app does not disallow DoH/DoT, all you need to do is point the custom DNS to the resolver on your machine that would then forward the DNS requests onto the DoH/DoT resolvers. On Linux, you could also add extra config to the tunnel interface via systemd-resolve to use DoH/DoT resolvers on the interface via the CLI - this obviously isn't great,  but it is a solution. We are planning on adding an option for our client to not try and manage DNS, in which case everything should work just fine.

**Comment by Noir16** [comment]
You could try Portmaster.
https://github.com/safing/portmaster

Android already offers a Private DNS feature, so use that.

**Comment by steel-steven-87** [comment]
I wish Mullvad implemented DNS over HTTPS support. After I learned about this feature, I really want it and it seems that the only privacy respecting VPN that has this at the moment is IVPN.

**Comment by twisty-thrifty205729** [comment]
Is this something that’s a possibility? I tried using Mullvad with the `ctrld` CLI and even after adding its bootstrap DNS server as a secondary custom DNS resolver so Mullvad would stop blocking it, it’s very unstable when Mullvad is connected (not sure whether that’s a Mullvad issue or a `ctrld` issue) so it’d be nice to be able to take it out of the equation.

**Comment by BionicBison05** [comment]
@pinkisemils now that the app contains an encrypted DNS resolver (I'm on the iOS beta), is this something that could be reasonably implemented in the near(ish) future?

**Comment by muhasturk** [comment]
Two years have passed and there's still no progress on this? Even the simplest VPN apps (such as Adguard VPN) have DOH support. How many more years will it take to get this done?

**Comment by faern** [comment]
It is unclear to us at Mullvad how this would improve privacy or security, or what your use case is that is currently not possible. If some of you who want this feature better explain their use-case that would be awesome. Without a clear motivation for **why** this feature is needed, we don't have a good reason to add the technical complexity that the feature would inevitably cost.

When you are connected to Mullvad, all your DNS requests go through the tunnel and are therefore encrypted by the tunnel. So what benefit does it add to encrypt the DNS request itself?
* Is it because you want to hide the DNS request content from Mullvad? Since the relay can see where you are connecting right after the DNS lookup, it does not add that much value really.
* Is it because you want to use a third party DNS service, that does not offer plain DNS? That's sort of a legit reason I guess.

When you are disconnected from Mullvad, the custom DNS is not used anyway.

**Comment by aniqueta** [comment]
My use case is being able to use Mullvad and, at the same time, my personal Pi-hole with my custom block and allow list, which also has private DNS entries. 

I could VPN into my own network, but then the exit IP is personally identifiable, and I don't get the benefits of Mullvad. And when traveling, the performance is dependent on distance to the VPN. It doesn't benefit from all the servers Mullvad runs. 

Thanks for considering.

**Comment by faern** [comment]
So you are out and about (somewhere in the world) and you want to run your traffic through Mullvad. But you want your DNS to go back home to your pi-hole? And you also feel like the DNS requests need to be encrypted between the Mullvad exit relay and your pi-hole (otherwise plain DNS to the Pi-hole could solve your problem)?

I see. That's a use case :+1: It could be somewhat solved by allowing custom blocklists directly in our app. Then you would not need to route DNS to your own home at all. Saving both time and adding some privacy. That would however not solve the problem of your private DNS entries.

**Comment by aniqueta** [comment]
Exactly! 

Exit to home DNS in the clear seems silly if one is privacy conscious. 

I'm wondering how difficult this is on at least Apple platforms. Mobile configuration profiles with the DoH or apps like DNSecure can already provide DNS while a VPN is running if the VPN uses 0.0.0.0 as the DNS and doesn't tunnel that traffic. I made a feature request fit this here, since this is how I workaround it right now using WireGuard directly instead. I'd prefer to use the Mullvad app, for simplicity and also to change locations to nearest when traveling. 

https://github.com/mullvad/mullvadvpn-app/issues/6203

**Comment by muhasturk** [comment]
>     * Is it because you want to use a third party DNS service, that does not offer plain DNS? That's sort of a legit reason I guess.

My motivation for wanting DNS over HTTPS is that I have a private DNS server and my own custom filters. When connected to a VPN, I want to use my own DNS server at the same time. While this is possible with Adguard, it is not possible with Mullvad.

**Comment by ghost** [comment]
ADD THIS ALREADY!!!!

**Comment by Rawa** [comment]
Just doing some chores here, we've moved to discussions. I'm going to close this issue and consolidate it into the very similar one for DoH & DoT https://github.com/mullvad/mullvadvpn-app/discussions/9121
